Computed properties that contain state-changing functions trigger vue/no-side-effects-in-computed-properties

<!--
  Before posting the issue, please confirm that the problem you're getting
  is not related with your code editor configuration.
  To make sure it's not, run: yarn eslint src/your-file.vue
-->

**Tell us about your environment**	
* **ESLint version:**	5.16.0
* **eslint-plugin-vue version:**	4.7.1
* **Node version:**	9.2.1

**Please show your full configuration:**
<!-- Paste content of your .eslintrc file -->
```js
module.exports = {
  root: true,
  env: {
    node: true,
  },
  extends: [
    'plugin:vue/essential',
    '@vue/airbnb',
  ],
  rules: {
    'no-console': process.env.NODE_ENV === 'production' ? 'error' : 'off',
    'no-debugger': process.env.NODE_ENV === 'production' ? 'error' : 'off',
    'no-param-reassign': 'off',
    'prefer-destructuring': 'off',
  },
  parserOptions: {
    parser: 'babel-eslint',
  },
};
```

**What did you do?**	
<!-- Please include the actual source code causing the issue. -->
```vue
<template>
  <div>
    <div
      v-for="tab in tabs"
      :key="tab.id"
      v-text="tab.text"
      :class="{ active: tab.active }"
      @click.prevent="tab.click"
    />
  </div>
</template>

<script>
export default {
  computed: {
    tabs() {
      return this.groups.map((group, i) => ({
        id: i,
        text: group.name,
        active: i === this.selectedTab,
        click: () => {
          // the below line is underlined, (vue/no-side-effects-in-computed-properties)
          this.selectedTab = i;
        },
      }));
    },
    groups() {
      return this.unfilteredGroups.filter(group => group.visible);
    },
  },

  data() {
    return {
      selectedTab: 0,
      unfilteredGroups: [
        { name: 'foo', visible: true },
        { name: 'bar', visible: true },
        { name: 'foobar', visible: false },
      ],
    };
  },
};
</script>

<style scoped>
.active {
  background-color: black;
  color: white;
  font-size: 18px;
  font-family: 'Comic Sans MS', monospace;
}
</style>
```

**What did you expect to happen?**

I expected to get no ESLint errors. In my opinion, this computed property does not have any side-effects - it only returns a value and does not change any state by itself.

**What actually happened?**
<!-- Please include the actual, raw output from ESLint. -->

```
/home/sean/projects/frontend/src/App.vue
  23:11  error  Unexpected side effect in "tabs" computed property  vue/no-side-effects-in-computed-properties

✖ 1 problem (1 error, 0 warnings)
```

--- a/lib/rules/no-side-effects-in-computed-properties.js
+++ b/lib/rules/no-side-effects-in-computed-properties.js
@@ -24,20 +24,38 @@ module.exports = {
 
   create (context) {
     const forbiddenNodes = []
+    let scopeStack = { upper: null, body: null }
+
+    function onFunctionEnter (node) {
+      scopeStack = { upper: scopeStack, body: node.body }
+    }
+
+    function onFunctionExit () {
+      scopeStack = scopeStack.upper
+    }
 
     return Object.assign({},
       {
+        ':function': onFunctionEnter,
+        ':function:exit': onFunctionExit,
+
         // this.xxx <=|+=|-=>
         'AssignmentExpression' (node) {
           if (node.left.type !== 'MemberExpression') return
           if (utils.parseMemberExpression(node.left)[0] === 'this') {
-            forbiddenNodes.push(node)
+            forbiddenNodes.push({
+              node,
+              targetBody: scopeStack.body
+            })
           }
         },
         // this.xxx <++|-->
         'UpdateExpression > MemberExpression' (node) {
           if (utils.parseMemberExpression(node)[0] === 'this') {
-            forbiddenNodes.push(node)
+            forbiddenNodes.push({
+              node,
+              targetBody: scopeStack.body
+            })
           }
         },
         // this.xxx.func()
@@ -46,7 +64,10 @@ module.exports = {
           const MUTATION_REGEX = /(this.)((?!(concat|slice|map|filter)\().)[^\)]*((push|pop|shift|unshift|reverse|splice|sort|copyWithin|fill)\()/g
 
           if (MUTATION_REGEX.test(code)) {
-            forbiddenNodes.push(node)
+            forbiddenNodes.push({
+              node,
+              targetBody: scopeStack.body
+            })
           }
         }
       },
@@ -54,11 +75,12 @@ module.exports = {
         const computedProperties = utils.getComputedProperties(obj)
 
         computedProperties.forEach(cp => {
-          forbiddenNodes.forEach(node => {
+          forbiddenNodes.forEach(({ node, targetBody }) => {
             if (
               cp.value &&
               node.loc.start.line >= cp.value.loc.start.line &&
-              node.loc.end.line <= cp.value.loc.end.line
+              node.loc.end.line <= cp.value.loc.end.line &&
+              targetBody === cp.value
             ) {
               context.report({
                 node: node,

--- a/lib/utils/index.js
+++ b/lib/utils/index.js
@@ -822,7 +822,7 @@ module.exports = {
         isFunc = true
       } else {
         if (n.computed) {
-          parsedCallee.push('[]')
+          parsedCallee.push('[]' + (isFunc ? '()' : ''))
         } else if (n.property.type === 'Identifier') {
           parsedCallee.push(n.property.name + (isFunc ? '()' : ''))
         }

--- a/tests/lib/rules/no-side-effects-in-computed-properties.js
+++ b/tests/lib/rules/no-side-effects-in-computed-properties.js
@@ -143,6 +143,36 @@ ruleTester.run('no-side-effects-in-computed-properties', rule, {
         }
       })`,
       parserOptions
+    },
+    {
+      code: `Vue.component('test', {
+        computed: {
+          test () {
+            return {
+              action1() {
+                this.something++
+              },
+              action2() {
+                this.something = 1
+              },
+              action3() {
+                this.something.reverse()
+              }
+            }
+          },
+        }
+      })`,
+      parserOptions
+    },
+    {
+      code: `Vue.component('test', {
+        computed: {
+          test () {
+            return this.something['a']().reverse()
+          },
+        }
+      })`,
+      parserOptions
     }
   ],
   invalid: [

